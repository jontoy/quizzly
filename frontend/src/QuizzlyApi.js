import axios from "axios";
import config from "./config";
const DATABASE_URL =
  process.env.REACT_APP_BASE_URL ||
  config.DATABASE_URL ||
  "http://localhost:3001";

class QuizzlyApi {
  static async request(endpoint, paramsOrData = {}, verb = "get") {
    console.debug("API Call:", endpoint, paramsOrData, verb);

    try {
      return (
        await axios({
          method: verb,
          url: `${DATABASE_URL}/${endpoint}`,
          [verb === "get" ? "params" : "data"]: paramsOrData,
        })
      ).data;
      // axios sends query string data via the "params" key,
      // and request body data via the "data" key,
      // so the key we need depends on the HTTP verb
    } catch (err) {
      console.error("API Error:", err.response);
      let message = err.response.data.message;
      throw Array.isArray(message) ? message : [message];
    }
  }
  static async getQuizzes({ name, min_difficulty, max_difficulty }) {
    let res = await this.request(`quiz`, {
      name,
      min_difficulty,
      max_difficulty,
    });
    return res.quizzes;
  }
  static async getQuiz(id) {
    let res = await this.request(`quiz/${id}`);
    return res.quiz;
  }
  static async getQuizAnswers(id) {
    let res = await this.request(`quiz/${id}/answers`);
    return res.answers;
  }
  static async getQuestions({ text, quiz_id }) {
    let res = await this.request(`question`, {
      text,
      quiz_id,
    });
    return res.quizzes;
  }
  static async getQuestionsByQuizId(quiz_id) {
    return await this.getQuestions({ quiz_id });
  }
  static async getQuestion(question_id) {
    let res = await this.request(`question/${question_id}`);
    return res.question;
  }
  static async getOptions({ value, question_id, is_correct }) {
    let res = await this.request(`option`, {
      value,
      question_id,
      is_correct,
    });
    return res.quizzes;
  }
  static async getOptionsByQuestionId(question_id) {
    return await this.getOptions({ question_id });
  }
  static async getOption(option_id) {
    let res = await this.request(`option/${option_id}`);
    return res.option;
  }
}

export default QuizzlyApi;
