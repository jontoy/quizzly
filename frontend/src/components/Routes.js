import React from "react";
import { Switch, Route, Redirect } from "react-router-dom";
import Question from "./Question";
import Home from "./Home";
import Quizzes from "./Quizzes";
import Quiz from "./Quiz";
import QuizResult from "./QuizResult";

const Routes = () => {
  return (
    <Switch>
      <Route exact path="/">
        <Home />
      </Route>
      <Route exact path="/quizzes/:quizId/questions/:questionNumber">
        <Question />
      </Route>
      <Route exact path="/quizzes/:quizId/results">
        <QuizResult />
      </Route>
      <Route exact path="/quizzes/:quizId">
        <Quiz />
      </Route>
      <Route exact path="/quizzes">
        <Quizzes />
      </Route>
      <Redirect to="/" />
    </Switch>
  );
};

export default Routes;
