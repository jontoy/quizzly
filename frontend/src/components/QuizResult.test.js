import React from "react";
import { render } from "@testing-library/react";
import { Router } from "react-router-dom";
import { Provider } from "react-redux";
import QuizzlyApi from "../QuizzlyApi";
import { createMemoryHistory } from "history";
import configureStore from "redux-mock-store";
import Routes from "./Routes";
const mockStore = configureStore([]);
jest.mock("../QuizzlyApi");

const quiz = {
  id: "abcdefg",
  name: "A quiz",
  difficulty: 6,
  questions: [
    { question_id: "q1id", text: "q1" },
    { question_id: "q2id", text: "q2" },
  ],
};

const answers = {
  id: "abcdefg",
  answers: [
    { question_id: "q1id", text: "q1", valid_options: ["opt1id"] },
    { question_id: "q2id", text: "q2", valid_options: ["opt2id"] },
  ],
};
QuizzlyApi.getQuiz.mockResolvedValue(quiz);
QuizzlyApi.getQuizAnswers.mockResolvedValue(answers);

describe("QuizResult tests", () => {
  let store;
  let component;
  beforeEach(() => {
    store = mockStore({
      quizzes: { [quiz.id]: quiz },
      responses: { [quiz.id]: ["opt1id", "opt3id"] },
    });
    const history = createMemoryHistory();
    history.push("/quizzes/abcdefg/results");
    component = render(
      <Provider store={store}>
        <Router history={history}>
          <Routes />
        </Router>
      </Provider>
    );
  });
  it("renders without crashing", () => {});
  it("matches snapshot", () => {
    expect(component.asFragment()).toMatchSnapshot();
  });
});
