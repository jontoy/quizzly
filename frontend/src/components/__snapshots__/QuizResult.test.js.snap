// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QuizResult tests matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="QuizResult container-fluid mt-3"
  >
    <div
      class="ResultsBanner jumbotron bg-info text-light"
    >
      <h1
        class="display-4"
      >
        So close!
      </h1>
      <p
        class="lead"
      >
        You got 1 out of 2 correct on this attempt.
      </p>
      <hr
        class="my-4"
      />
      <p>
        You almost had it!
      </p>
    </div>
    <div
      class="mt-4"
    >
      <ul
        class="list-group list-group-flush"
      >
        <li
          class="list-group-item list-group-item-secondary"
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-check fa-w-16 text-success"
            data-icon="check"
            data-prefix="fas"
            focusable="false"
            role="img"
            viewBox="0 0 512 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M173.898 439.404l-166.4-166.4c-9.997-9.997-9.997-26.206 0-36.204l36.203-36.204c9.997-9.998 26.207-9.998 36.204 0L192 312.69 432.095 72.596c9.997-9.997 26.207-9.997 36.204 0l36.203 36.204c9.997 9.997 9.997 26.206 0 36.204l-294.4 294.401c-9.998 9.997-26.207 9.997-36.204-.001z"
              fill="currentColor"
            />
          </svg>
           What you know
        </li>
        <li
          class="list-group-item text-black-50"
        >
          q1
        </li>
      </ul>
    </div>
    <div
      class="mt-4"
    >
      <ul
        class="list-group list-group-flush"
      >
        <li
          class="list-group-item list-group-item-secondary"
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-times fa-w-11 text-danger"
            data-icon="times"
            data-prefix="fas"
            focusable="false"
            role="img"
            viewBox="0 0 352 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M242.72 256l100.07-100.07c12.28-12.28 12.28-32.19 0-44.48l-22.24-22.24c-12.28-12.28-32.19-12.28-44.48 0L176 189.28 75.93 89.21c-12.28-12.28-32.19-12.28-44.48 0L9.21 111.45c-12.28 12.28-12.28 32.19 0 44.48L109.28 256 9.21 356.07c-12.28 12.28-12.28 32.19 0 44.48l22.24 22.24c12.28 12.28 32.2 12.28 44.48 0L176 322.72l100.07 100.07c12.28 12.28 32.2 12.28 44.48 0l22.24-22.24c12.28-12.28 12.28-32.19 0-44.48L242.72 256z"
              fill="currentColor"
            />
          </svg>
           What you should review
        </li>
        <li
          class="list-group-item text-black-50"
        >
          q2
        </li>
      </ul>
    </div>
    <div
      class="container-fluid bg-light py-2"
    >
      <button
        class="btn btn-outline-danger"
        role="button"
      >
        Retry Quiz?
      </button>
      <a
        href="/quizzes"
      >
        <button
          class="btn btn-outline-info ml-1"
        >
          Find more Quizzes
        </button>
      </a>
    </div>
  </div>
</DocumentFragment>
`;
